enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,483 @@
+
+>Note: Best effort will be made to adhere to the UX mentioned above however appropriate fot this command too. 
+
+##### Introducing a new `validate` subcommand for `opm index`
+
+A new subcommand `validate` will be introduced under `opm index`, that will validate the content of each package's json representation. These validations will be in the context of rules about the structure of an index (required fields, name conventions etc of bundles, packages, and the index itself). A best effort is currently being made to aggregate these rules from various sources and maintained in the [api project's validation library](https://github.com/operator-framework/api/tree/master/pkg/validation). The `opm index validate` will be reponsible for validating the index for authors to make sure that the index conforms to the rules mentioned above. 
+
+#### Package cofig CUE definition
+
+CUE(https://cuelang.org/docs/about/) is a configuration and data validation language that makes it easy to define the structure of a resource and validate it's content. This section will discuss CUE as a possible implementation to define the structure of packages inside an index, and use that definition to generate the json representation of the packages, and ultimately the representation of the index. 
+
+##### Schema Defintion
+
+An example definition of a package that lives in the [api](https://github.com/operator-framework/api) can be: 
+
+```bash=
+$ cat package.cue
+#gvk: {
+    group: string
+    version: string
+    kind: string
+}
+
+#bundle: {
+    path: string
+    version: string
+    channel: string
+    upgradeMode: "semver"|"semver-skippatch"
+    label?: string
+    ownedAPIs: [...#gvk]
+    requiredAPIs: [...#gvk]
+}
+
+#package: {
+    name: string
+    defaultChannel: string
+    icon: string
+    bundles: [...#bundle]
+}
+```
+
+##### Generating JSON representation using schema defintion
+
+When a bundle is added to an index using `opm`, `opm` will first make a copy of `pacakage.cue` locally, and then check if the package mentioned in the bundle metadata already exists in the index. If it does not, `opm` will append the corresponding `bundle` and `package` CUE structs to the end of the local copy: 
+
+For example, when `opm index add --bundle=quay.io/mynamespace/amq-streamsv1.0.0@channel:beta --tag=docker.io/my-namespace/community-operators` is called, the end of the file will look like so: 
+
+```bash=
+$ tail -28 local.copy.package.cue
+amqstreamsv100: #bundle & {
+    path: "quay.io/mynamespace/amq-streamsv1.0.0"
+    version: "v1.0.0"
+    channel: "beta"
+    upgradeMode: "semver" 
+    ownedAPIs: [
+        {
+            group: "amq-streams.org.com"
+            version: "v1beta1"
+            kind: "amq-streams"
+        },    
+    ]
+    requiredAPIs: [
+        {
+            group: "etcd.database.coreos.com"
+            version: "v1alpha1"
+            kind: "etcd"
+        },    
```

Comment:
Why are API requirements / ownership part of the catalog definition?

Reply:
Answered [here](https://github.com/operator-framework/enhancements/pull/37#issuecomment-742844778)